homeassistant/components/nest/config_flow.py

```diff
@@ -107,58 +150,124 @@ def extra_authorize_data(self) -> Dict[str, str]:
-        # Update existing config entry when in the reauth flow.  This
-        # integration only supports one config entry so remove any prior entries
-        # added before the "single_instance_allowed" check was added
+    @property
+    def _defaults(self) -> dict:
+        """Return default values from existing entry or configuration.yaml."""
+        if self._reauth_data and CONF_PROJECT_ID in self._reauth_data:
+            return self._reauth_data
+        return self.hass.data.get(DOMAIN, {}).get(DATA_NEST_CONFIG, {})
+
+    async def _cleanup_invalid_config_entries(self):
+        """Update or remove non-conforming config entries."""
         existing_entries = self.hass.config_entries.async_entries(DOMAIN)
         if existing_entries:
-            updated = False
-            for entry in existing_entries:
-                if updated:
-                    await self.hass.config_entries.async_remove(entry.entry_id)
-                    continue
-                updated = True
+            # Add a unique_id if not present on existing config entry
+            first_entry = existing_entries[0]
+            if first_entry.unique_id != DOMAIN:
                 self.hass.config_entries.async_update_entry(
-                    entry, data=data, unique_id=DOMAIN
+                    first_entry, data=first_entry.data, unique_id=DOMAIN
                 )
-                await self.hass.config_entries.async_reload(entry.entry_id)
+            # Remove entries added before the "single_instance_allowed" check
+            # was implemented in the config flow
+            for entry in existing_entries[1:]:
+                await self.hass.config_entries.async_remove(entry.entry_id)
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user, and dispatch between APIs."""
+        if not self.is_sdm_api():
+            return await self.async_step_init(user_input)
+        # This method async_step_user overrides the first step of the OAuth
+        # flow to insert a step (async_step_device_access) to prompt for
+        # necessary OAuth related configuration options.  The parent class
+        # async_step_user is then invoked to actually initiate the OAuth flow.
+        await self._cleanup_invalid_config_entries()
+        existing_entry = await self.async_set_unique_id(DOMAIN)
+        if existing_entry and not self._reauth_data:
+            return self.async_abort(reason="single_instance_allowed")
+        return await self.async_step_device_access()
+
+    async def async_step_device_access(self, user_input=None):
+        """Enter configuration options for the Device Access project."""
+        if user_input is None:
+            data_schema = {
+                vol.Required(
+                    CONF_PROJECT_ID, default=self._defaults.get(CONF_PROJECT_ID)
+                ): str,
+                vol.Required(
+                    CONF_CLIENT_ID, default=self._defaults.get(CONF_CLIENT_ID)
```

Comment:
Why should we make client id and secret configurable via config flow? That's not in line with our design standards.

Reply:
Ah my mistake.   Have a pointer to history or docs I should brush up on?

My intention here is to flow this:
https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision

I interpret this as meaning we shouldn't be using configuration.yaml to communicate with nest devices.  Is OAuth an exception or am I misunderstanding? Or something that we'd like to change but it needs more thought?

Reply:
Yeah, oauth credentials is an exception. We don't want that user facing. Preferably we want to add account linking for a majority of oauth supporting integrations so the user don't need to add any credentials in yaml.

Reply:
OK great.  I am really motivated to make life simpler for users setting up this integration. 

Would it be appropriate for me to start tracking this down?   When we were updating the docs for the oddities of redirect uris this came up in a discussion with @balloob and @frenck on discord "So we need to maybe prioritize our account linking service to work with local client ID/secret" -- I would be happy to help push this forward, if that is appropriate.  I see https://developers.google.com/assistant/identity/oauth-concept-guide and related pages seem to have some useful resources that I can start digging through.

I'm still a bit unclear on *why* OAuth client id and secret are an exception and why `configuration.yaml` is currently preferred, so would be great to understand more just so I have the context.    I get the sense that neither are preferred and its just a matter of making it automatic if possible.

Note that in the case of nest integration, users still have to go configure an oauth client id anyway and let the nest device access console know about it, as part of setup.  Not having to specify it one more place seems nice though.

Reply:
Oauth credentials is an exception cause we don't want users to have to configure this. It's an advanced concept. Config flows should be easy for users to use. Config yaml is for advanced users.

Open an architecture issue for implementation advice and design direction.

Reply:
Thanks, I will move this to a draft, do some research, then try to start the architecture discussion.

That is, my view is that a hyrbd `configuration.yaml` + GUI config flow for the nest integration would *not* be an improvement, so moving to config flow should probably block on figuring that out.

Meanwhile, happy for folks to chime in with any other directional feedback or prior art / discussions.

Reply:
The idea is that we want the user to opt-in if they want to do their own manual flow. Otherwise once we sort out account linking, we don't know if we should show both options because the user could always add their own client ID/secret (although 99% wouldn't use it)

Right now it's YAML-based but I think ideally we would have a separate page where users can configure client ID/secret to use for integrations.